contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,564 @@
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-db
+spec:
+    containers:
+    - name: mysql
+      image: mysql
+      volumeMounts:
+      - name: my-db-data
+	mountPath: /var/lib/mysql/data
+    volumes:
+    - name: my-db-data
+      persistentVolumeClaim:
+	claimName: local-raw-pvc
+```
+  NOTE: *accessModes correspond to the container runtime values. Where RWO == RWM (mknod) to enable the device to be written to and
+  create new files. (Default is RWM) ROX == R
+  **(RWX is NOT valid for block and should return an error.)** * This has been validated among runc, Docker and rocket. 
+
+
+
+## UC4: 
+
+DESCRIPTION: StorageClass with non-dynamically created volumes
+
+BACKGROUND: The admin wishes to create a storage class that will identify pre-provisioned block PVs based on a user's PVC request for volumeType: Block. 
+
+WORKFLOW: 
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: block-volume
+provisioner: no-provisioning 
+parameters:
+  volumeType: block
+```
+* Sample of pre-created volume definition:
+
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+ name: pv-block-volume
+spec:
+ volumeType: block
+ storageClassName: block-volume
+ capacity:
+   storage: 35Gi
+ accessModes:
+   - ReadWriteOnce
+ local:
+    path: /dev/xvdc
+```
+## [FUTURE] UC5: 
+
+DESCRIPTION: StorageClass with dynamically created volumes 
+
+BACKGROUND: The admin wishes to create a storage class that will dynamically create block PVs based on a user's PVC request for volumeType: Block. The admin desires the volumes be created dynamically and deleted when the PV definition is deleted. 
+
+WORKFLOW:
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: local-fast
+provisioner: kubernetes.io/local-block-ssd
+parameters:
+  volumeType: block
+```
+
+***This has implementation details that have yet to be determined. It is included in this proposal for completeness of design ****
+
+## UC6: 
+
+DESCRIPTION: The developer wishes to request a block device via a Storage Class.
+
+WORKFLOW:
+
+USER: 
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+ name: pvc-local-block
+spec:
+ volumeType: block
+ storageClassName: local-fast
+ accessModes:
+  - ReadWriteOnce
+ resources:
+   requests:
+     storage: 10Gi
+```  
+
+## UC7:
+
+DESCRIPTION: User wishes to install a software defined storage solution  (i.e.GlusterFS/Ceph(bluestore)) on top of raw block device. Using something like a StateFulSet for GlusterFS, for example, the system could provide it with provisioned block storage, and allow the software defined storage solution to consume it as needed. 
+
+BACKGROUND: This provides a way to dynamically create raw block devices to later be consumed by software defined storage.
+
+WORKFLOW:
+
+ADMIN:
+
+```
+kind: StorageClass
+apiVersion: storage.k8s.io/v1beta1
+metadata:
+  name: aws-ebs-raw
+provisioner: kubernetes.io/aws-ebs-raw
+parameters:
+  type: gp2
+  fstype: raw 
```

Comment:
@jsafrane this was a suggestion by @lpabon that they use at CoreOS, I will remove. Ultimately it can be anything the provisioner wants it to be and can understand